solvers/python_pgd.py

```diff
@@ -22,29 +36,38 @@ class Solver(BaseSolver):
     ]
 
     def skip(self, X, y, lmbd, fit_intercept):
+        if sparse.issparse(X) and self.use_gpu:
+            return True, "sparse is not supported with GPU"
+
         # XXX - not implemented but not too complicated to implement
         if fit_intercept:
             return True, f"{self.name} does not handle fit_intercept"
 
         return False, None
 
     def set_objective(self, X, y, lmbd, fit_intercept):
+        if self.use_gpu:
+            # transfert X, y to GPU
+            X, y = cp.array(X), cp.array(y)
+
         self.X, self.y, self.lmbd = X, y, lmbd
         self.fit_intercept = fit_intercept
 
     def run(self, callback):
         L = self.compute_lipschitz_constant()
 
+        xp = cp if self.use_gpu else np
+
         n_features = self.X.shape[1]
-        w = np.zeros(n_features)
+        w = xp.zeros(n_features)
         if self.use_acceleration:
-            z = np.zeros(n_features)
+            z = xp.zeros(n_features)
 
         t_new = 1
         while callback(w):
```

Comment:
This raises an error in the GPU case.
``w`` is on the GPU when calling the callback. However, operations aren't permitted between CPU and GPU variables.

One solution is to transfer back ``w`` to CPU before calling callbacks yet I'm afraid this deteriorates the timing benchmarks of the solver as data transfer takes (often) time.

Reply:
Indeed this may be an issue. But compared to the copy of X the overhead here should be neglectable, it's a single vector ?

Otherwise switch to another stopping strategy so that the transfer is performed only once, and the cost will be amortized for enough iterations.

Reply:
I think we better switch the stopping criterion to iteration